prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -179,6 +192,121 @@ def filter_by_connected_components(lines, substations):
+        if voltages.max() > state_threshold_100_161 * voltages.sum()
+        else None
+        for state, voltages in state_100_161_lines.items()
+    }
+    lines.loc[missing_mask & class_100_161_mask, "VOLTAGE"] = lines.loc[
+        missing_mask & class_100_161_mask
+    ].STATE_1.map(most_common_100_161_voltage)
+    lines.drop("STATE_1", axis=1, inplace=True)
+
+    # Create a mapping of lines to neighboring lines
+    sub_1_groupby = lines.groupby("SUB_1_ID")
+    sub_2_groupby = lines.groupby("SUB_2_ID")
+    empty_df = pd.DataFrame()
+    neighbors = lines.loc[lines.VOLTAGE.isna()].apply(
+        lambda x: set().union(
+            safe_get_group(sub_1_groupby, x.SUB_1_ID, empty_df).index,
+            safe_get_group(sub_1_groupby, x.SUB_2_ID, empty_df).index,
+            safe_get_group(sub_2_groupby, x.SUB_1_ID, empty_df).index,
+            safe_get_group(sub_2_groupby, x.SUB_2_ID, empty_df).index,
+        ),
+        axis=1,
+    )
+
+    # Update lines with missing voltage if only one voltage exists among their neighbors
+    # Iteratively update until no more updates are possible
+    while True:
+        num_missing = len(lines.query("VOLTAGE.isnull()"))
+        if num_missing == 0:
+            break
+        neighbor_voltages = lines.loc[lines.VOLTAGE.isna()].apply(
+            lambda x: {
+                v
+                for v in set(lines.loc[neighbors.loc[x.name], "VOLTAGE"])
+                if v == v  # this eliminates float("nan") values from the set
+            },
+            axis=1,
+        )
+        single_voltages = neighbor_voltages.apply(
+            lambda x: next(iter(x)) if len(x) == 1 else None
+        )
+        if single_voltages.isna().all():
+            print(f"{num_missing} line voltages can't be found via neighbor consensus")
+            break
+        lines.loc[
+            lines.VOLTAGE.isna() & ~single_voltages.isna(), "VOLTAGE"
+        ] = single_voltages.loc[~single_voltages.isna()]
+
+    # Update lines with missing voltage using lowest voltage among their neighbors
+    # Iteratively update until no more updates are possible
+    while True:
+        num_missing = len(lines.query("VOLTAGE.isnull()"))
+        if num_missing == 0:
+            break
+        neighbor_voltages = lines.loc[lines.VOLTAGE.isna()].apply(
+            lambda x: {
+                v
+                for v in set(lines.loc[neighbors.loc[x.name], "VOLTAGE"])
+                if v == v  # this eliminates float("nan") values from the set
+            },
+            axis=1,
+        )
+        lowest_voltages = neighbor_voltages.apply(
+            lambda x: min(x) if len(x) > 0 else None
+        )
+        if lowest_voltages.isna().all():
+            print(f"{num_missing} line voltages can't be found via lowest_neighbor")
+            break
+        lines.loc[
+            lines.VOLTAGE.isna() & ~lowest_voltages.isna(), "VOLTAGE"
+        ] = lowest_voltages.loc[~lowest_voltages.isna()]
```

Comment:
Maybe we can write a function that returns _a set of neighbor voltages of a given line_, which could be used in both of the two layers here?

Reply:
That would definitely be cleaner.

Reply:
I went one step further and generalized a function that iterates over the neighboring line voltages of missing-voltage lines, and does _something_ to those voltages to decide which voltages to update, printing a message and stopping once no more voltages can be updated.